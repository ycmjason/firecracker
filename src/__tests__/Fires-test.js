import { when } from "jest-when";

jest.mock("../FiresCollection");
import {
  // eslint-disable-next-line no-unused-vars
  Fires,
  FiresCollection,
  FiresDocument
} from "..";

const createMockFirestore = (base = {}) => {
  return {
    _config: {
      settings: { timestampsInSnapshots: true }
    },
    ...base
  };
};

describe("Fires", () => {
  it("constructor should throw if Firestore did not set timestampsInSnapshots", () => {
    const $mockFirestore = {
      _config: {
        settings: { timestampsInSnapshots: false }
      }
    };
    expect(() => new Fires($mockFirestore)).toThrowErrorMatchingInlineSnapshot(
      `"Fires: Firestore must set \`timestampsInSnapshots\` to \`true\`."`
    );
  });

  describe("fires.collection(name)", () => {
    it("should return a FiresCollection", () => {
      const $mockFirestore = createMockFirestore({ collection: jest.fn() });

      when($mockFirestore.collection)
        .calledWith("mockCollectionName")
        .mockReturnValue("$mockCollection");

      FiresCollection.mockImplementation($collection => {
        if ($collection === "$mockCollection") {
          return { type: "collection" };
        }
        return {};
      });

      const db = new Fires($mockFirestore);
      expect(db.collection("mockCollectionName")).toEqual({
        type: "collection"
      });
    });

    it("should return the same FiresCollection on subsequent calls", () => {
      const $mockFirestore = createMockFirestore({ collection: jest.fn() });

      when($mockFirestore.collection)
        .calledWith("mockCollectionName")
        .mockReturnValue("$mockCollection");

      FiresCollection.mockImplementation($collection => {
        if ($collection === "$mockCollection") {
          return { type: "collection" };
        }
        return {};
      });

      const db = new Fires($mockFirestore);
      const collection1 = db.collection("mockCollectionName");
      const collection2 = db.collection("mockCollectionName");
      const collection3 = db.collection("mockCollectionName");
      expect(collection1).toBe(collection2);
      expect(collection2).toBe(collection3);
    });
  });
});
