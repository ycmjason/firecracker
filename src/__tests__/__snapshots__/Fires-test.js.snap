// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fires constructor should throw if Firestore did not set timestampsInSnapshots 1`] = `"Fires: Firestore must set \`timestampsInSnapshots\` to \`true\`. Please add this line before initiating Fires \`firebase.firestore().settings({ timestampsInSnapshots: true });\`"`;
